cEP-0017.md

```diff
@@ -0,0 +1,356 @@
+
+### Ansible for downloading bear and dependencies
+
+1. Write a script in python to create a playbook that will have the installation instruction
+   for all bears and their dependencies.
+   - The script goes over all bears and loops over its `REQUIREMENT`. The cases for different
+     tasks that will be generated in the playbook for different kinds of `REQUIREMENT`
+     are shown in the code samples.
+   - Additionally the script would add relevant tags to each installation task. This would
+     include the bear name, language etc.
+   - Other options for user interaction can also supported by Ansible and can be easily added.
+
+# Code Samples
+
+## cib improvements
+
+### Making use of conda packages in cib
+
+```python
+    def install_bears(bear_names_list):
+        bear_failed_list = []
+        for bear_name in bear_names_list:
+            if PipRequirement(bear_name).is_installed() or
+               CondaRequirement(bear_name).is_installed():
+                    print(bear_name + ' is already installed.')
+                    continue
+            if not PipRequirement(bear_name).install_package() or
+               not CondaRequirement(bear_name).install_package():
+                    install_requirements(bear_name)
+                    install_bear_dependencies(bear_name)
+                    print('DONE')
+            else:
+                print('Failed')
+                bears_failed_list.append(bear_name)
+        return bears_failed_list
+```
+
+### Installing bear dependencies
+
+```python
+    def install_bear_dependencies(bear_name):
+        package_object = importlib.import_module(
+            'coala' + package_name + '.' + package_name)
+        for bear in getattr(package_object, bear_name).BEAR_DEPS:
+            install_bears(bear)
+```
+
+### UI to ask permission for dependencies to be installed
+
+```python
+    import coalib.misc.Constants as consts
+
+    def display_dependencies(bears):
+        print('The bears have the following dependencies')
+        for bear in bears:
+            for requirement in bear.REQUIREMENTS:
+                print(requirement.package, requirement.version)
+
+        choice = raw_input().lower()
+        if  choice  in  consts.TRUE_STRINGS:
+            return   True
+        elif  choice  in  consts.FALSE_STRINGS:
+            return   False
+        else :
+            sys.stdout.write( "Please respond with 'yes' or 'no'" )
+```
+
+### Splitting `DistributionRequirement`
+
+```
+from coala_utils.decorators import generate_eq, generate_repr
+from sarge import run, Capture
+
+
+@generate_eq("platform", "package", "version", "repo")
+@generate_repr()
+class PlatformRequirement:
+    """
+    This class helps keeping track of bear distribution requirements.
+    """
+
+    REQUIREMENTS = {}
+
+    def __init__(self, platform: str, package: str, version="", repo=""):
+        """
+        :param platform:  A string with the name of the distribution
```

Comment:
platform needs to be a controlled list, and that list needs to be documented in this cEP.

It is the essence of this project.

After defining the list, also identify which are required for this project, and which are maybes, and which are out of scope.

BSDs are definitely out of scope, as that is the other projects focus.